### PR TITLE
Only set title when ConPTY is unavailable.

### PR DESCRIPTION
**Pull request by drichardson**
The window title is used by a non-ConPTY hack to save/restore the window in some situations.

This PR changes that behavior to only set the title when required to implement the non-ConPTY hack. On newer implementations where ConPTY is available, the window title is now left alone.

Addresses point (1) in PowerShell/Win32-OpenSSH#1309, and also renames some functions/variables to make it obvious in the code that this hack should not be used for ConPTY consoles.

### CODE REVIEW

**Comment by DHowett** [comment]
THANK YOU

**Comment by DHowett** [comment]
(speaking as the owner of the console.)

**Comment by bagajjal** [comment]
@DHowett  - What is the win32-API to find the current application console window instead of using window title.
If we have such then it solves two problems,
1) We still have problem when users have multiple console windows using ssh.
2) I would like to set the console title all the time no matter it's conpty / non-conpty.

Thank you.

**Comment by DHowett** [comment]
@bagajjal sure!

1. `GetConsoleWindow`. It returns the HWND, and has existed for 35 years. There is no reason to ever get the console window -- why do you need it still?
2. `SetConsoleTitle`. Why do you want to set the title? Ideally, the title will be provided by the EXE file or by the LNK the user has created.


If the user makes a `.lnk` to ssh.exe like `my cool server.lnk`, the default title will be "my cool server". If you override the title, you will delete their title.

**Comment by bagajjal** [comment]
@DHowett - I want to have consistent behavior.
Instead of using FindWindow(), we can use GetConsoleWindow() and avoid setting the title.

**Comment by DHowett** [comment]
Yes, your behavior will be wrong but at least it will be _consistent_. Do you have a functional spec formalizing the need for this feature? :smile:

**Comment by DHowett** [comment]
I must reiterate, though: There is no reason to get the console HWND, and _there is no guarantee that the console is hosted in a window._ The HWND you get back will be meaningless if you are running inside conpty (like, inside Windows Terminal).

File: contrib/win32/win32compat/console.c [hunk 1]
```diff
@@ -36,6 +36,7 @@
 #include <windows.h>
 #include <conio.h>
 #include <io.h>
+#include <assert.h>
```
Comment (bagajjal) [inline contrib/win32/win32compat/console.c:39]: Please remove asserts.

File: contrib/win32/win32compat/console.c [hunk 2]
```diff
@@ -75,7 +78,7 @@ typedef struct _SCREEN_RECORD {
 
 PSCREEN_RECORD pSavedScreenRec = NULL;
 int in_raw_mode = 0;
-char *consoleTitle = "OpenSSH SSH client";
+static const char *consoleTitle_NonConPTYHack = "OpenSSH SSH client";
```
Comment (bagajjal) [inline contrib/win32/win32compat/console.c:81]: Replace FindWindow() with GetConsoleWindow() at line 1612

Comment (bagajjal) [reply contrib/win32/win32compat/console.c:81]: We can remove this line (81) and lines 146-149.

File: contrib/win32/win32compat/console.c [hunk 3]
```diff
@@ -140,7 +143,10 @@ ConEnterRawMode()
 		return;
 	}
 
-	SetConsoleTitle(consoleTitle);
+	/* Only overwrite the console title if ConPTY is not supported. This is
```
Comment (bagajjal) [inline contrib/win32/win32compat/console.c:146]: We can remove this after replacing FindWindow() with GetConsoleWindow() API

File: contrib/win32/win32compat/console.c [hunk 4]
```diff
@@ -1589,18 +1595,21 @@ ConSaveScreen()
 }
 
 void
-ConSaveViewRect()
+ConSaveViewRect_NonConPTYHack()
 {
+	assert(!is_conpty_supported());
```
Comment (bagajjal) [inline contrib/win32/win32compat/console.c:1600]: Please remove asserts

File: contrib/win32/win32compat/console.c [hunk 5]
```diff
@@ -1589,18 +1595,21 @@ ConSaveScreen()
...
 	CONSOLE_SCREEN_BUFFER_INFO csbi;
 	if (GetConsoleScreenBufferInfo(GetConsoleOutputHandle(), &csbi))
 		SavedViewRect = csbi.srWindow;
 }
 
 void
-ConRestoreViewRect()
+ConRestoreViewRect_NonConPTYHack()
 {
+	assert(!is_conpty_supported());
```
Comment (bagajjal) [inline contrib/win32/win32compat/console.c:1609]: Please remove asserts

File: contrib/win32/win32compat/console.c [hunk 6]
```diff
@@ -1589,18 +1595,21 @@ ConSaveScreen()
...
+
 	CONSOLE_SCREEN_BUFFER_INFO consoleInfo;
-	HWND hwnd = FindWindow(NULL, consoleTitle);
+	HWND hwnd = FindWindow(NULL, consoleTitle_NonConPTYHack);
```
Comment (bagajjal) [inline contrib/win32/win32compat/console.c:1612]: Replace with GetConsoleWindow()

**Comment by bagajjal** [comment]
@drichardson - I had offline chat with @DHowett to clarify his comments. He is good with replacing FindWindow() with GetConsoleWindow(). I left few comments, please fix them.

**Comment by drichardson** [comment]
> 
> 
> @drichardson - I had offline chat with @DHowett to clarify his comments. He is good with replacing FindWindow() with GetConsoleWindow(). I left few comments, please fix them.

@bagajjal I'm not using PowerShell openssh-portable, I switched to MSYS2's ssh, so someone else will need to do the work if you want to take this upstream.

**Comment by bagajjal** [comment]
> > @drichardson - I had offline chat with @DHowett to clarify his comments. He is good with replacing FindWindow() with GetConsoleWindow(). I left few comments, please fix them.
> 
> @bagajjal I'm not using PowerShell openssh-portable, I switched to MSYS2's ssh, so someone else will need to do the work if you want to take this upstream.

Thanks @drichardson . I'm closing this PR. 
I will fix this as part of https://github.com/PowerShell/openssh-portable/pull/494

**Pull request closed without merging**